sql/analyzer/optimization_rules.go

```diff
@@ -327,6 +327,11 @@ func simplifyFilters(ctx *sql.Context, a *Analyzer, node sql.Node, scope *Scope,
 
 				return e, transform.SameTree, nil
 			case *expression.Like:
+				// if the charset is utf8mb3, the last character used in optimization rule does not work
+				charset := getCharsetInUse(ctx, e.Left)
```

Comment:
We have a function for getting the collation of an expression.
https://github.com/dolthub/go-mysql-server/blob/main/sql/expression/collatedexpr.go#L133
You can ignore the returned integer and grab the character set from the collation. This will also handle the case where no collation is specified.